Source/MQTTnet/Internal/AsyncReadWriteLock.cs

```diff
@@ -0,0 +1,214 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace MQTTnet.Internal
+{
+    public class AsyncReadWriteLock : IDisposable
+    {
+        private readonly object _syncRoot = new object();
+        private readonly Queue<ReadWriteLockWaiter> _readWaiters = new Queue<ReadWriteLockWaiter>(64);
+        private readonly Queue<ReadWriteLockWaiter> _writeWaiters = new Queue<ReadWriteLockWaiter>();
+        private readonly IDisposable _readReleaser;
+        private readonly IDisposable _writeReleaser;
+        private readonly Task<IDisposable> _readCompletedTask;
+        private readonly Task<IDisposable> _writeCompletedTask;
+        private int _readLockCount;
+        private bool _isLockedForWrite;
+        private bool _isDisposed;
+
+        public AsyncReadWriteLock()
```

Comment:
Please add multiple unit tests and cover all features of this lock so that we can proof it is working.

Reply:
Added multiple unit tests.